src/com/amazon/ion/benchmark/WriteRandomIonValues.java

```diff
@@ -271,111 +294,127 @@ public static void writeRandomSymbolValues(int size, String format, String path)
         int count = 0;
         // Determine how many values should be written before the writer.flush(), and this process aims to reduce the execution time of writer.flush().
         while (currentSize <= 0.05 * size) {
-            WriteRandomIonValues.writeDataToFile(writer, constraintStruct);
+            WriteRandomIonValues.writeDataToFile(writer, parsedTypeDefinition);
             count += 1;
             writer.flush();
             currentSize = (int) file.length();
         }
         while (currentSize <= size) {
             for (int i = 0; i < count; i++) {
-                WriteRandomIonValues.writeDataToFile(writer, constraintStruct);
+                WriteRandomIonValues.writeDataToFile(writer, parsedTypeDefinition);
             }
             writer.flush();
             currentSize = (int) file.length();
         }
     }
 
     /**
-     * This method will be reused by different data generator
-     * @param writer writer is IonWriter.
-     * @param constraintStruct is an IonStruct which contains the top-level constraints in Ion Schema.
-     * @throws IOException if an error occur during the data writing process.
+     * Generating data which is conformed with provided constraints and writing it to the output file.
+     * @param writer is IonWriter.
+     * @param parsedTypeDefinition is parsed from ion schema file as IonStruct format, it contains the top-level constraints.
+     * @throws Exception if an error occur during the data writing process.
      */
-    private static void writeDataToFile(IonWriter writer, IonStruct constraintStruct) throws Exception {
-        IonType type = IonType.valueOf(constraintStruct.get(IonSchemaUtilities.KEYWORD_TYPE).toString().toUpperCase());
-        IonValue value = null;
-        // Check whether the 'valid_values' constraints provided in the top level constraint struct.
-        // If a list of 'valid_values' provided, the generated value should be selected randomly from the provided 'valid_values' list every iteration.
-        // Constraint 'valid_values' has three formats. <RANGE<TIMESTAMP>>, <RANGE<NUMBER>> and [ <VALUE>... ]. This step only check the format [ <VALUE>... ].
-        // If the annotation 'range' has been detected, it will be processed in the constructing data steps.
-        if (constraintStruct != null) {
-            value = IonSchemaUtilities.parseValidValues(constraintStruct);
-        }
-        if (value != null && IonSchemaUtilities.getConstraintValueAsRange(constraintStruct, IonSchemaUtilities.KEYWORD_VALID_VALUES) == null) {
-            value.writeTo(writer);
+    private static void writeDataToFile(IonWriter writer, ReparsedType parsedTypeDefinition) throws Exception {
+        // The first step is to check whether parsedTypeDefinition contains 'valid_values'. The reason we prioritize checking
+        // 'valid_values' is that the constraint 'type' might not be contained in the type definition, in that case we cannot trigger
```

Comment:
If any type definition doesn't contain a type constraint its assumed to be default `type:any` constraint. So this could be applied to [logical constraints](https://amzn.github.io/ion-schema/docs/spec.html#logic-constraints) as well.
e.g.
```
type::{
   name: "my_type",
   one_of: [string, int],
   type: any <--- default type `any`
}
```

```
type::{
   name: "my_type",
   valid_values: ["a", 5, 4.5],
   type: any <--- default type `any`
}
```

Definition for type [any](https://amzn.github.io/ion-schema/docs/spec.html#core-types): represents any of the types listed above as well as types derived from those types

Reply:
> If any type definition doesn't contain a type constraint its assumed to be default `type:any` constraint. So this could be applied to [logical constraints](https://amzn.github.io/ion-schema/docs/spec.html#logic-constraints) as well. e.g.
> 
> ```
> type::{
>    name: "my_type",
>    one_of: [string, int],
>    type: any <--- default type `any`
> }
> ```
> 
> ```
> type::{
>    name: "my_type",
>    valid_values: ["a", 5, 4.5],
>    type: any <--- default type `any`
> }
> ```
> 
> Definition for type [any](https://amzn.github.io/ion-schema/docs/spec.html#core-types): represents any of the types listed above as well as types derived from those types

Thanks for correction. The assumption of 'constraint `type` is required when constraint `valid_values` is not provided' is not aligned with ion schema specification grammar. The logic of getting data type information will refer to this [comment]( https://github.com/amzn/ion-java-benchmark-cli/pull/34#discussion_r889203146). This change will be updated in the next PR.